### DISCUSSION TITLE
`dict` default key type should be `Hashable`, not `Any`

### DISCUSSION THREAD

**Discussion by randolf-scholz** [python/typing#1248] [General]
Python dictionaries raise a runtime-error when fed a non-hashable object. Therefore, the default key-type of a blank `dict` type hint should be `Hashable` not `Any`

- `dict` ⇝ `dict[Hashable, Any]`

**Comment by erictraut** [comment]
Unspecified type arguments default to `Any` for all generic types. This is well established in PEP 484.

While it is allowed to leave type arguments unspecified, it is not recommended. In other words, you generally shouldn't use a "bare" `dict` in a type annotation. If your intent is `dict[Hashable, Any]`, then you should spell that out.

There is a [draft PEP](https://peps.python.org/pep-0696/) that proposes to add support for default type arguments, but even with this new facility, it wouldn't make sense to default the first type parameter of `dict` to `Hashable` because the type parameters for `dict` are invariant.

**Comment by jeremyn** [comment]
Just curious, but when would this matter? mypy doesn't report any problem on line 3 of the following code for me:

```python
d: dict = {}
for k in d:
    hash(k)
```
